metaquot/expander/ppx_metaquot_expander.ml

```diff
@@ -236,74 +230,73 @@ module Patt = struct
     | _ ->
       Location.raise_errorf ?loc:(loc_of_extension ext)
         "pattern expected"
-  let convert = Conversion.ast_to_pattern
   let extension entry = Ppx_bootstrap.Extension.Patt entry
 end
 
 let expected loc string = raise (Ppx_bootstrap.Expected.Expected (loc, string))
 
+let single_signature_item_payload payload =
+  match Payload.to_concrete payload with
+  | None | Some (PStr _ | PTyp _ | PPat _) -> None
+  | Some (PSig signature) ->
+    match Signature.to_concrete signature with
+    | None | Some [] | Some (_ :: _ :: _) -> None
+    | Some [item] -> Some item
+
+let single_structure_item_payload payload =
+  match Payload.to_concrete payload with
+  | None | Some (PSig _ | PTyp _ | PPat _) -> None
+  | Some (PStr structure) ->
+    match Structure.to_concrete structure with
+    | None | Some [] | Some (_ :: _ :: _) -> None
+    | Some [item] -> Some item
+
+let single_expression_payload payload =
+  match single_structure_item_payload payload with
+  | None -> None
+  | Some item ->
+    match Structure_item.to_concrete item with
+    | None -> None
+    | Some item ->
+      match Structure_item_desc.to_concrete item.pstr_desc with
+      | Some Pstr_eval (e, attr) -> Some (e, attr)
+      | Some _ | None -> None
```

Comment:
BTW, we should eventually put these in the standard ppx_view combinators. /cc @NathanReb

Reply:
Isn't that going to be Ast_pattern's job in the end? I would assume that only ppx_view and metaquot have to deal with this and that other PPX authors would simply use `Extension.declare` and `Ast_pattern` to describe the shape of the payload.

I'm totally okay with adding such views but unless I'm missing something, they probably won't be used much!

Reply:
I'm not entirely sure how the `Extension.declare` API should evolve. i.e. once we have `match%view`, does it still make sense to describe the extension payload using the pattern functions or should we just write a plain `%view` pattern?

Reply:
What I like about the current API is that error reporting will be consistent accross different ppx-es when the payload doesn't fit the right shape. I agree in most cases the shape of the payload is too complex to be described using `Ast_pattern` only and that some of the error reporting then falls into the ppx author's responsibility but at least they all share that common root.

It might not be justified but I feel more comfortable adding arbitrary combinators to something like `Ast_pattern` than to `Ppx_ast.Viewer`. The main reason for that is the resulting code would look like:
```ocaml
match%view payload with
| Single_expr_payload expr -> ...
| PPat pat -> ...
```
And I feel like this might be very confusing to the untrained eye, more confusing than using a documented function from Ast_pattern. This would allow mixing combinators that fit the exact AST types and completely arbitrary ones (although helpful). If you're not used to ppx_view and how it expands pattern matching into function calls, it might be hard finding out where this function comes from or where and how to find and use them in the first place.

That might just be solved by having proper documentation and carefully splitting the generated parts (that match the AST) from the handwritten ones such as `single_expr_payload` so the documentation is easy to browse and it's easy to discover them.